### PR TITLE
[ffmpeg] Add support for file, h264 options

### PR DESCRIPTION
**Pull request by qmfrederik**
This PR adds support for two new FFmpeg configuration options: file and h264.

### CODE REVIEW

**Comment by Osyotr** [comment]
Vcpkg features should be additive. Is `ffmpeg[core]` not slim enough for you?

File: ports/ffmpeg/vcpkg.json [hunk 1]
```diff
@@ -344,6 +344,9 @@
         "dav1d"
       ]
     },
+    "disable-everything": {
+      "description": "Disable all encoders, decoders, hwaccels, muxers, demuxers, parsers, bitstream filters, protocols, input devices, output devices and filters, unless explicitely enabled."
+    },
```
Comment (dg0yt) [inline ports/ffmpeg/vcpkg.json:349]: AFAIU this feature isn't additive. "Features must be treated as additive functionality."
https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#do-not-use-features-to-implement-alternatives
And there is `vcpkg install ffmpeg[core]`.

**Review by dg0yt — COMMENTED** [review]
Declaring new features in the manifest usually doesn't have any effect on the build, without changing `portfile.cmake`.

**Comment by qmfrederik** [comment]
Thanks @Osyotr and @dg0yt. Looks like ffmpeg[core] ships with a copy of avutil which is about 200KB larger than what you get with --disable-everything, which is fine.

I've updated the PR to drop the disable-everything option and keep file and h264 only.  Let me know what you think. Thanks!

File: ports/ffmpeg/vcpkg.json [hunk 2]
```diff
@@ -431,6 +434,9 @@
     "gpl": {
       "description": "Allow use of GPL code, the resulting libs and binaries will be under GPL"
     },
+    "h264": {
```
Comment (Osyotr) [inline ports/ffmpeg/vcpkg.json:437]: No dependencies?

File: ports/ffmpeg/vcpkg.json [hunk 3]
```diff
@@ -410,6 +410,9 @@
         }
       ]
     },
+    "file": {
```
Comment (Osyotr) [inline ports/ffmpeg/vcpkg.json:413]: This is very generic.
Also should be in `default-features`.

File: ports/ffmpeg/portfile.cmake [hunk 4]
```diff
@@ -310,6 +310,10 @@ else()
     set(OPTIONS "${OPTIONS} --disable-libfdk-aac")
 endif()
 
+if("file" IN_LIST FEATURES)
+    set(OPTIONS "${OPTIONS} --enable-protocol=file")
```
Comment (Osyotr) [inline ports/ffmpeg/portfile.cmake:314]: No `--disable...`?

File: ports/ffmpeg/portfile.cmake [hunk 5]
```diff
@@ -328,6 +332,10 @@ else()
     set(OPTIONS "${OPTIONS} --disable-libfribidi")
 endif()
 
+if("h264" IN_LIST FEATURES)
+    set(OPTIONS "${OPTIONS} --enable-decoder=h264 --enable-parser=h264 --enable-demuxer=h264")
```
Comment (Osyotr) [inline ports/ffmpeg/portfile.cmake:336]: No `--disable...`?

**Comment by qmfrederik** [comment]
So, I compared `ffmpeg[core,h264]` with `--disable-everything --enable-decoder=h264`

| library | core | disable-everything |
|-|-|
| avcodec-58 | 11,9 MB | 1,68 MB
| avfilter-8 | 3,38 MB | -
| avformat-59 | 2,10 MB | 180 KB 
| avutil-57 | 705 KB | 518 KB
| swscale-5 | 602 KB | 494 KB

So that's almost 15 MB extra.

**Comment by qmfrederik** [comment]
So, in the end, what worked for me was to add options to enable/disable the default components which make up the various FFmpeg libraries for at least the core build.  I've amended the PR, let me know what you think.

**Comment by qmfrederik** [comment]
@Osyotr @dg0yt I've reworked the PR based on your feedback, thoughts?

**Comment by Sibras** [comment]
This PR doesnt do anything as file and h264 are already enabled in default ffmpeg configuration. So ffmpeg[core] already has all of these features enabled so adding extra command line parameters to enable them individually doesnt do anything.

I can see your original intention was to disable everything then only enable the small file and h264 components. But using specific options for fine-grained enablement of configs is massive as ffmpeg has 1000s of similar small pieces that can be enabled or disabled using its own configuration system. Vcpkg has deliberetly avoided that as its huge and ever changing (so keeping it in sync with source is difficult). So even if these components werent enabled as default im not sure we want to support adding just 2 specific options for one particular use case given we cant support all of them and then others may want similar options added for their own specific use cases.

**Review by BillyONeal — CHANGES_REQUESTED** [review]
I agree with @Osyotr

**Comment by qmfrederik** [comment]
Closing for now, will revisit later if time allows.

**Pull request closed without merging**